flang/lib/Lower/CallInterface.cpp

```diff
@@ -181,28 +181,57 @@ static inline auto AsGenericExpr(T e) {
   return Fortran::evaluate::AsGenericExpr(Fortran::common::Clone(e));
 }
 
-mlir::Value Fortran::lower::CallerInterface::getResultLength() {
-  // FIXME: technically, this is a specification expression,
-  // so it should be evaluated on entry of the region we are
-  // in, it can go wrong if the specification expression
-  // uses a symbol that may have changed.
-  //
-  // The characteristic has to be explicit for such
-  // cases, so the allocation could also be handled on callee side, in such
-  // case. For now, protect with an unreachable.
+void Fortran::lower::CallerInterface::walkResultLengths(
+    ExprVisitor visitor) const {
   assert(characteristic && "characteristic was not computed");
-  const auto *typeAndShape =
-      characteristic->functionResult.value().GetTypeAndShape();
+  const auto &result = characteristic->functionResult.value();
+  const auto *typeAndShape = result.GetTypeAndShape();
   assert(typeAndShape && "no result type");
-  auto expr = AsGenericExpr(typeAndShape->LEN().value());
-  if (Fortran::evaluate::IsConstantExpr(expr)) {
-    Fortran::lower::StatementContext stmtCtx;
-    auto exv = converter.genExprValue(expr, stmtCtx);
-    assert(!stmtCtx.hasCleanups());
-    return fir::getBase(exv);
+  auto dynamicType = typeAndShape->type();
+  if (dynamicType.category() == Fortran::common::TypeCategory::Character) {
+    if (!dynamicType.charLength()->isDeferred())
+      visitor(AsGenericExpr(typeAndShape->LEN().value()));
+  } else if (dynamicType.category() == common::TypeCategory::Derived) {
+    const auto &derivedTypeSpec = dynamicType.GetDerivedTypeSpec();
+    if (Fortran::semantics::CountLenParameters(derivedTypeSpec) > 0)
+      TODO(converter.genLocation(),
+           "function result with derived type length parameters");
   }
-  llvm_unreachable(
-      "non constant result length on caller side not yet safely handled");
+}
+
+void Fortran::lower::CallerInterface::walkResultExtents(
+    ExprVisitor visitor) const {
+  using Attr = Fortran::evaluate::characteristics::TypeAndShape::Attr;
+  assert(characteristic && "characteristic was not computed");
+  const auto &result = characteristic->functionResult.value();
+  const auto *typeAndShape = result.GetTypeAndShape();
+  assert(typeAndShape && "no result type");
+  if (typeAndShape->attrs().test(Attr::DeferredShape))
+    return;
+  for (auto extent : typeAndShape->shape())
+    visitor(AsGenericExpr(extent.value()));
+}
+
+bool Fortran::lower::CallerInterface::mustMapInterfaceSymbols() const {
```

Comment:
What is the rational of MLIR for avoiding const member functions ? Putting const on member functions that should not modify the object seem a very sane things to me. Isn't it because MLIR classes achieve data protection by other means ?

Reply:
There was some paper on it in the MLIR documentation way back. Looking for it, I instead found https://mlir.llvm.org/docs/Rationale/UsageOfConst/ which looks like it is more focused with the IR itself. At any rate, the question was really more whether we needed const correctness here or whether access controls would be sufficient.